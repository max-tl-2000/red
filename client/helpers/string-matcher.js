/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import React from 'react';
import Caption from 'components/Typography/Caption';
import trim from 'helpers/trim';

export const renderStringWithMatch = ({ value, className, bold, inline, ellipsis, searchQueryValue, id, TagElement = Caption }) => {
  value = trim(value);
  searchQueryValue = trim(searchQueryValue);

  if (value.toLowerCase().includes(searchQueryValue.toLowerCase())) {
    const startHighlightIndex = value.toLowerCase().indexOf(searchQueryValue.toLowerCase());
    const endHighlightIndex = startHighlightIndex + searchQueryValue.length;

    return (
      <TagElement className={className} bold={bold} inline={inline} ellipsis={ellipsis} key={id}>
        {value.slice(0, startHighlightIndex)}
        <TagElement bold inline highlight>
          {value.slice(startHighlightIndex, endHighlightIndex)}
        </TagElement>
        {value.slice(endHighlightIndex, value.length)}
      </TagElement>
    );
  }
  return (
    <TagElement className={className} bold={bold} inline={inline} ellipsis={ellipsis} key={id}>
      {value}
    </TagElement>
  );
};
