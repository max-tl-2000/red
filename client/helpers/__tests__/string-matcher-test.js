/*
 * Copyright (c) 2022 Reva Technology Inc., all rights reserved.
 * Unauthorized copying of this file, via any medium is strictly prohibited
 * Licensed under the Elastic License 2.0; you may not use this file except
 * in compliance with the Elastic License 2.0.
 */

import React from 'react';
import { mount } from 'enzyme';
import { mountToJson } from 'enzyme-to-json';
import { renderStringWithMatch } from '../string-matcher';

describe('renderStringWithMatch', () => {
  describe('when a null value is provided', () => {
    it('should render the list and not throw', () => {
      const props = {
        value: null,
        className: 'myClass',
        bold: true,
        inline: false,
        ellipsis: false,
        searchQueryValue: 'Some value',
        id: '1',
      };

      const mountPoint = mount(<div>{renderStringWithMatch(props)}</div>);
      expect(mountToJson(mountPoint)).toMatchSnapshot();
    });
  });

  describe('when a non null value is provided and a searchQueryValue is provided', () => {
    it('should highlight the matches', () => {
      const props = {
        value: 'Peter Parker',
        className: 'myClass',
        bold: true,
        inline: false,
        ellipsis: false,
        searchQueryValue: 'Pet',
        id: '1',
      };

      const mountPoint = mount(<div>{renderStringWithMatch(props)}</div>);
      expect(mountToJson(mountPoint)).toMatchSnapshot();
    });
  });

  describe('when searchQuery is not found in value', () => {
    it('should render the string without highlight matches', () => {
      const props = {
        value: 'Bruce wayne',
        className: 'myClass',
        bold: true,
        inline: false,
        ellipsis: false,
        searchQueryValue: 'Pet',
        id: '1',
      };

      const mountPoint = mount(<div>{renderStringWithMatch(props)}</div>);
      expect(mountToJson(mountPoint)).toMatchSnapshot();
    });
  });
});
