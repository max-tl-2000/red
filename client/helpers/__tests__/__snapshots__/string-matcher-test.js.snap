// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderStringWithMatch when a non null value is provided and a searchQueryValue is provided should highlight the matches 1`] = `
<div>
  <caption
    bold={true}
    className="myClass"
    ellipsis={false}
    inline={false}
  >
    <p
      className="caption bold myClass"
      data-component="caption"
      data-typography-element={true}
    >
      <caption
        bold={true}
        highlight={true}
        inline={true}
      >
        <span
          className="caption bold highlight"
          data-component="caption"
          data-typography-element={true}
        >
          Pet
        </span>
      </caption>
      er Parker
    </p>
  </caption>
</div>
`;

exports[`renderStringWithMatch when a null value is provided should render the list and not throw 1`] = `
<div>
  <caption
    bold={true}
    className="myClass"
    ellipsis={false}
    inline={false}
  >
    <p
      className="caption bold myClass"
      data-component="caption"
      data-typography-element={true}
    />
  </caption>
</div>
`;

exports[`renderStringWithMatch when searchQuery is not found in value should render the string without highlight matches 1`] = `
<div>
  <caption
    bold={true}
    className="myClass"
    ellipsis={false}
    inline={false}
  >
    <p
      className="caption bold myClass"
      data-component="caption"
      data-typography-element={true}
    >
      Bruce wayne
    </p>
  </caption>
</div>
`;
